### ISSUE TITLE
Question about Callouts

### ISSUE THREAD

**Issue by skostojohn** [smukov/async-apex-framework#2]
Thanks for this great framework! 
I have been playing around with it a bit, and I've hit a snag.
I want to make a callout from one of my handlers; however, this is not possible as AsyncRequestQueueable creates a savepoint prior to invoking the handler.  This counts as a DML operation, which must come after any callouts.  Consequently the platform won't allow me to callout from my handler.

Any ideas for an elegant solution?

Thanks!!

Scott

**Comment by skostojohn** [comment]
I think I have a solution....
I removed the savepoint and rollback from AsyncRequestQueueable, and in the catch I fire a platform event to write the error message onto the request.  Then I throw the exception manually and let the platform handle rolling back the transaction.  The error message persists since platform events are not rolled back.

Thank you!
Scott

**Comment by smukov** [comment]
Hi Scott,

That’s a pretty nice solution and a great use of Platform Events. I'm just wondering, are you also re-starting the `AsyncRequestQueueable` from the Platform Event trigger? 

I'm asking because if you're not catching the exception anymore the transaction will be rolled back and the remaining `AsyncRequestQueueable` code will not execute, so the next request will not be processed either until something else re-starts the queueable.

Thanks,
Miroslav

**Comment by skostojohn** [comment]
Thanks Miroslav.  I am not firing `AsyncRequestQueueable` from the platform event trigger - I currently have plenty of queueable transactions going so I don't think firing again from the platform event trigger is necessary.

However today I am fighting a much bigger problem with this framework.  I am seeing multiple instances of AsyncRequestQueueable processing the same Async_Request__c record.  I don't think this should not be possible, because the first instance should lock the record (due to the use of FOR UPDATE) and subsequent instances should not be able to get it, but that is not what I am seeing.

I added some more logging with platform events, and it is clear that multiple different jobIDs (instances of `AsyncRequestQueueable`) are processing the same request Ids.  I found this problem because I was getting unhandled exceptions from the delete of the request, saying that entity was already deleted (by one of the other instances)

Here is a screenshot of some of my logging:
<img width="846" alt="image" src="https://user-images.githubusercontent.com/2495345/60984009-d2db9480-a2ef-11e9-9195-8526e490d844.png">

I am not sure how to address this.  I have a lot of instances of `AsyncRequestQueueable` running (invoked from triggers) - maybe I switch to a scheduled job to fire them?  But the locking approach should work!!!

Any ideas?

Thanks

Scott
